src/qt/listctrl.cpp

```diff
@@ -13,125 +13,1020 @@
+
+        wxListEvent event;
+        InitListEvent(event,
+                      GetHandler(),
+                      wxEVT_LIST_END_LABEL_EDIT,
+                      current_index);
+        event.m_item.SetText(editedText);
+
+        if (hint == QAbstractItemDelegate::RevertModelCache)
+        {
+            event.SetEditCanceled(true);
+            EmitEvent(event);
+        }
+        else
+        {
+            // Allow event handlers to decide whether to accept edited text
+            if (!GetHandler()->HandleWindowEvent(event) || event.IsAllowed())
+                m_itemDelegate.AcceptModelData(editor, model(), current_index);
+        }
+
+        // wx doesn't have hints to edit next/previous item
+        if ( hint == QAbstractItemDelegate::EditNextItem ||
+             hint == QAbstractItemDelegate::EditPreviousItem )
+        {
+            hint = QAbstractItemDelegate::SubmitModelCache;
+        }
+
+        QTreeView::closeEditor(editor, hint);
+        closePersistentEditor(current_index);
     }
 
     wxTextCtrl *GetEditControl()
     {
-        return m_editorFactory.GetEditControl();
+        return m_itemDelegate.GetEditControl();
     }
 
-private:
-    void itemClicked(QTreeWidgetItem * item, int column);
-    void itemActivated(QTreeWidgetItem * item, int column);
-    void itemPressed(QTreeWidgetItem * item, int column);
-
-    void ChangeEditorFactory()
+    virtual void paintEvent(QPaintEvent *event) wxOVERRIDE
     {
-        QItemDelegate *qItemDelegate = static_cast<QItemDelegate*>(itemDelegate());
-        qItemDelegate->itemEditorFactory();
-        qItemDelegate->setItemEditorFactory(&m_editorFactory);
+        QTreeView::paintEvent(event);
     }
 
-    wxQtTreeItemEditorFactory m_editorFactory;
+private:
+    void itemClicked(const QModelIndex &index);
+    void itemActivated(const QModelIndex &index);
+    void itemPressed(const QModelIndex &index);
+
+    wxQtStyledItemDelegate m_itemDelegate;
+    wxRecursionGuardFlag m_closingEditor;
 };
 
 wxQtListTreeWidget::wxQtListTreeWidget( wxWindow *parent, wxListCtrl *handler )
-    : wxQtEventSignalHandler< QTreeWidget, wxListCtrl >( parent, handler ),
-      m_editorFactory(handler)
+    : wxQtEventSignalHandler< QTreeView, wxListCtrl >( parent, handler ),
+    m_itemDelegate(handler),
+    m_closingEditor(0)
 {
-    connect(this, &QTreeWidget::itemClicked, this, &wxQtListTreeWidget::itemClicked);
-    connect(this, &QTreeWidget::itemPressed, this, &wxQtListTreeWidget::itemPressed);
-    connect(this, &QTreeWidget::itemActivated, this, &wxQtListTreeWidget::itemActivated);
+    connect(this, &QTreeView::clicked, this, &wxQtListTreeWidget::itemClicked);
+    connect(this, &QTreeView::pressed, this, &wxQtListTreeWidget::itemPressed);
+    connect(this, &QTreeView::activated, this, &wxQtListTreeWidget::itemActivated);
 
-    ChangeEditorFactory();
+    setItemDelegate(&m_itemDelegate);
+    setEditTriggers(NoEditTriggers);
```

Comment:
Why do we do this? Shouldn't we allow at least some triggers if `wxLC_EDIT_LABELS` is set?

Reply:
I believe this was to prevent you from being able to edit any cell in the list, when normally you can only edit the first column. This was an issue for us as we were suddenly able to start an edit of read only data in our list view. We are still able to edit our labels using the wxLC_EDIT_LABELS flag, I think there's just no way of starting an edit using one of the Qt triggers (Clicking selected, double clicking, edit key etc.). I believe this matches the MSW listctrl sample behaviour, if there is some way of starting an edit through mouse interaction I don't know of a way we could implement in Qt and prevent it from applying to all cells.